File: backend/app/schemas/conflict_aware_generation.py
from datetime import date
from enum import StrEnum
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


class OptimizationStatus(StrEnum):
    COMPLETE = "complete"
    IMPROVED_PARTIAL = "improved_partial"
    UNCHANGED = "unchanged"
    FAILED = "failed"
    STALE = "stale"


class BlockingReasonCode(StrEnum):
    LECTURER_OCCUPIED = "LECTURER_OCCUPIED"
    ROOM_OCCUPIED = "ROOM_OCCUPIED"
    COHORT_OCCUPIED = "COHORT_OCCUPIED"
    LECTURER_UNAVAILABLE = "LECTURER_UNAVAILABLE"
    ROOM_UNAVAILABLE = "ROOM_UNAVAILABLE"
    NO_ELIGIBLE_LECTURER = "NO_ELIGIBLE_LECTURER"
    NO_ELIGIBLE_ROOM = "NO_ELIGIBLE_ROOM"
    INSUFFICIENT_ROOM_CAPACITY = "INSUFFICIENT_ROOM_CAPACITY"
    UNAVAILABLE_DATE = "UNAVAILABLE_DATE"
    NO_ALLOWED_DATE_OR_WINDOW = "NO_ALLOWED_DATE_OR_WINDOW"
    COURSE_CONSTRAINT = "COURSE_CONSTRAINT"
    SELECTED_COURSE_COMPETITION = "SELECTED_COURSE_COMPETITION"
    INVALID_PLANNING_INPUT = "INVALID_PLANNING_INPUT"
    STALE_PLANNING_INPUT = "STALE_PLANNING_INPUT"


class OperationError(BaseModel):
    code: str
    message: str


class BlockingReason(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    code: BlockingReasonCode
    message: str
    related_count: int = Field(default=1, alias="relatedCount", ge=1)


class ArrangementImprovement(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    added_units: int = Field(alias="addedUnits", ge=0)
    reduced_conflicts: int = Field(alias="reducedConflicts", ge=0)
    reduced_lecturer_changes: int = Field(alias="reducedLecturerChanges", ge=0)
    reduced_room_changes: int = Field(alias="reducedRoomChanges", ge=0)


class OptimizationPreparationRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True, extra="forbid")

    semester_id: int = Field(alias="semesterId", ge=1)
    course_ids: list[int] = Field(alias="courseIds")
    unavailable_dates: list[date] = Field(default_factory=list, alias="unavailableDates")


class PreparedOptimizationCourse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    course_id: int = Field(alias="courseId")
    course_name: str | None = Field(alias="courseName")
    available: bool
    draft_schedule_id: int | None = Field(alias="draftScheduleId")
    draft_revision: int | None = Field(alias="draftRevision")
    scheduled_units: int = Field(alias="scheduledUnits", ge=0)
    remaining_units: int = Field(alias="remainingUnits", ge=0)
    replacement_required: bool = Field(alias="replacementRequired")
    input_snapshot_token: str = Field(alias="inputSnapshotToken", min_length=1)


class OptimizationPreparationResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    semester_id: int = Field(alias="semesterId")
    unavailable_dates: list[date] = Field(alias="unavailableDates")
    shared_snapshot_token: str = Field(alias="sharedSnapshotToken")
    courses: list[PreparedOptimizationCourse]
    replacement_course_ids: list[int] = Field(alias="replacementCourseIds")


class PreparedOptimizationCourseInput(BaseModel):
    model_config = ConfigDict(populate_by_name=True, extra="forbid")

    course_id: int = Field(alias="courseId", ge=1)
    expected_draft_schedule_id: int | None = Field(alias="expectedDraftScheduleId")
    expected_draft_revision: int | None = Field(alias="expectedDraftRevision", ge=1)
    input_snapshot_token: str = Field(alias="inputSnapshotToken", min_length=1)

    @model_validator(mode="after")
    def matching_draft_identity(self):
        if (self.expected_draft_schedule_id is None) != (self.expected_draft_revision is None):
            raise ValueError("Draft Schedule ID and revision must both be present or both be null.")
        return self


class OptimizationGenerationRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True, extra="forbid")

    semester_id: int = Field(alias="semesterId", ge=1)
    unavailable_dates: list[date] = Field(alias="unavailableDates")
    shared_snapshot_token: str = Field(alias="sharedSnapshotToken", min_length=1)
    replacement_confirmed: bool = Field(alias="replacementConfirmed")
    courses: list[PreparedOptimizationCourseInput]


class CourseOptimizationOutcome(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    course_id: int = Field(alias="courseId")
    course_name: str | None = Field(alias="courseName")
    status: OptimizationStatus
    draft_schedule_id: int | None = Field(alias="draftScheduleId")
    draft_revision: int | None = Field(alias="draftRevision")
    scheduled_units: int = Field(alias="scheduledUnits", ge=0)
    remaining_units: int = Field(alias="remainingUnits", ge=0)
    saved: bool
    improvement: ArrangementImprovement | None
    reasons: list[BlockingReason]
    errors: list[OperationError]


class OptimizationSummary(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    total: int = Field(ge=1, le=20)
    complete: int = Field(ge=0)
    improved_partial: int = Field(alias="improvedPartial", ge=0)
    unchanged: int = Field(ge=0)
    failed: int = Field(ge=0)
    stale: int = Field(ge=0)
    scheduled_units: int = Field(alias="scheduledUnits", ge=0)
    remaining_units: int = Field(alias="remainingUnits", ge=0)
    elapsed_milliseconds: int = Field(alias="elapsedMilliseconds", ge=0, le=60000)
    optimal_for_prepared_snapshot: bool = Field(alias="optimalForPreparedSnapshot")


class OptimizationGenerationResult(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    semester_id: int = Field(alias="semesterId")
    summary: OptimizationSummary
    outcomes: list[CourseOptimizationOutcome]


class ReplacementConfirmationRequired(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    code: Literal["REPLACEMENT_CONFIRMATION_REQUIRED"] = "REPLACEMENT_CONFIRMATION_REQUIRED"
    message: str
    replacement_course_ids: list[int] = Field(alias="replacementCourseIds")


class RequestFailureResponse(BaseModel):
    errors: list[OperationError]


class OptimizationOperationFailure(BaseModel):
    code: Literal[
        "OPTIMAL_RESULT_NOT_PROVEN",
        "OPTIMIZATION_MODEL_INVALID",
        "OPTIMIZATION_OPERATION_FAILED",
    ]
    message: str
    saved: Literal[False] = False

File: client/src/api/conflictAwareGeneration.ts
export type OptimizationStatus = 'complete' | 'improved_partial' | 'unchanged' | 'failed' | 'stale'

export type OptimizationError = {
  code: string
  message: string
  replacementCourseIds?: number[]
}

export type BlockingReason = {
  code: string
  message: string
  relatedCount: number
  holidayDate?: string
  holidayName?: string
  sourceKind?: 'teaching_session' | 'active_exam'
  sourceId?: number
}

export type PreparedOptimizationCourse = {
  courseId: number
  courseName: string | null
  available: boolean
  unavailableReasons: BlockingReason[]
  draftScheduleId: number | null
  draftRevision: number | null
  scheduledUnits: number
  remainingUnits: number
  replacementRequired: boolean
  effectiveConstraints: GenerationConstraints
  inputSnapshotToken: string
}

export type OptimizationPreparation = {
  semesterId: number
  scheduleRevisionId: number
  unavailableDates: string[]
  sharedSnapshotToken: string
  courses: PreparedOptimizationCourse[]
  replacementCourseIds: number[]
}

export type ArrangementImprovement = {
  addedUnits: number
  reducedConflicts: number
  reducedLecturerChanges: number
  reducedRoomChanges: number
}

export type CourseOptimizationOutcome = {
  courseId: number
  courseName: string | null
  status: OptimizationStatus
  draftScheduleId: number | null
  draftRevision: number | null
  scheduledUnits: number
  remainingUnits: number
  saved: boolean
  improvement: ArrangementImprovement | null
  reasons: BlockingReason[]
  errors: OptimizationError[]
}

export type OptimizationGenerationResult = {
  mode: 'direct_saved'
  semesterId: number
  summary: {
    total: number
    complete: number
    improvedPartial: number
    unchanged: number
    failed: number
    stale: number
    scheduledUnits: number
    remainingUnits: number
    elapsedMilliseconds: number
    optimalForPreparedSnapshot: boolean
  }
  outcomes: CourseOptimizationOutcome[]
}

export type CoverageFacts = {
  requiredUnits: number
  scheduledUnits: number
  remainingUnits: number
  status: 'complete' | 'partial'
}

export type ResolvedWarning = { code: string; count: number }

export type CourseRegenerationComparison = {
  courseId: number
  courseName: string | null
  current: CoverageFacts
  generated: CoverageFacts
  resolvedCurrentWarnings: ResolvedWarning[]
  remainingReasons: BlockingReason[]
}

export type RegenerationComparison = {
  selectedCourseIds: number[]
  current: CoverageFacts
  generated: CoverageFacts
  courses: CourseRegenerationComparison[]
  replacesAllSelectedSessions: true
  mayReplacePlannerEdits: true
}

export type PreparedGenerationEvidence = {
  semesterId: number
  scheduleRevisionId: number
  unavailableDates: string[]
  sharedSnapshotToken: string
  courses: {
    courseId: number
    expectedDraftScheduleId?: number | null
    expectedDraftRevision?: number | null
    inputSnapshotToken: string
  }[]
}

export type OptimizationDecisionRequiredResult = {
  mode: 'decision_required'
  saved: false
  candidateFingerprint: string
  preparedEvidence: PreparedGenerationEvidence
  comparison: RegenerationComparison
}

export type OptimizationGenerationResponse =
  | OptimizationGenerationResult
  | OptimizationDecisionRequiredResult

const API_BASE = import.meta.env.VITE_API_BASE_URL ?? ''

export async function prepareConflictAwareGeneration(
  semesterId: number,
  scheduleRevisionId: number,
  courseIds: number[],
  unavailableDates: string[] = [],
): Promise<OptimizationPreparation> {
  validateSelection(courseIds)
  const response = await request(`${API_BASE}/api/draft-schedules/optimization/prepare`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ semesterId, scheduleRevisionId, courseIds, unavailableDates }),
  })
  if (!response.ok) throw await parseOptimizationError(response)
  return response.json()
}

export async function generateConflictAwareSchedules(
  preparation: OptimizationPreparation,
): Promise<OptimizationGenerationResponse> {
  validateSelection(preparation.courses.map((course) => course.courseId))
  const response = await request(`${API_BASE}/api/draft-schedules/optimization/generate`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      semesterId: preparation.semesterId,
      scheduleRevisionId: preparation.scheduleRevisionId,
      unavailableDates: preparation.unavailableDates,
      sharedSnapshotToken: preparation.sharedSnapshotToken,
      courses: preparation.courses.map((course) => ({
        courseId: course.courseId,
        expectedDraftScheduleId: course.draftScheduleId,
        expectedDraftRevision: course.draftRevision,
        inputSnapshotToken: course.inputSnapshotToken,
      })),
    }),
  })
  if (!response.ok) throw await parseOptimizationError(response)
  return response.json()
}

export async function acceptConflictAwareSchedules(
  preview: OptimizationDecisionRequiredResult,
): Promise<OptimizationGenerationResult> {
  const evidence = preview.preparedEvidence
  validateSelection(evidence.courses.map((course) => course.courseId))
  const response = await request(`${API_BASE}/api/draft-schedules/optimization/accept`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      ...evidence,
      candidateFingerprint: preview.candidateFingerprint,
    }),
  })
  if (!response.ok) throw await parseOptimizationError(response)
  return response.json()
}

function validateSelection(courseIds: number[]) {
  if (courseIds.length < 1 || courseIds.length > 20) {
    throw [{ code: 'INVALID_OPTIMIZATION_SIZE', message: 'Semester optimization requires 1-20 courses.' }]
  }
  if (new Set(courseIds).size !== courseIds.length) {
    throw [{ code: 'DUPLICATE_COURSE_SELECTION', message: 'Select each course only once.' }]
  }
}

async function parseOptimizationError(response: Response): Promise<OptimizationError[]> {
  try {
    const body = await response.json()
    if (Array.isArray(body.errors)) return body.errors
    if (Array.isArray(body.detail)) {
      return body.detail.map((item: { msg?: string }) => ({ code: 'INVALID_REQUEST', message: item.msg ?? 'The optimization request is invalid.' }))
    }
    if (body.code) return [body]
  } catch {
    // Use the safe fallback below.
  }
  return [{ code: 'REQUEST_FAILED', message: 'Could not complete semester optimization.' }]
}

async function request(input: RequestInfo | URL, init?: RequestInit): Promise<Response> {
  try {
    return await fetch(input, init)
  } catch {
    throw [{ code: 'NETWORK_ERROR', message: 'Could not reach the backend API.' }]
  }
}
import type { GenerationConstraints } from './draftSchedule'
